### DISCUSSION TITLE
Make TypeArray deref into slices?

### DISCUSSION THREAD

**Discussion by DexterHill0** [wasm-bindgen/wasm-bindgen#3806] [Q&A] [answered]
### Summary

Maybe I'm missing something, but is there any way to take a typed array (say, UInt8Array) and turn it into a `&[u8]`, or the equivalent? I found issue #5 but that was a very long time ago and I'm not sure if its still applicable. 

In my use case I load a large image as an `ArrayBuffer` from an HTTP request, convert it into a `UInt8Array` and pass that to the WASM. Inside the WASM is a function that takes a `&[u8]` and it would be convinient, and more efficient, to be able to take a reference to the `UInt8Array` rather than first allocate an enourmous vector, then take a slice of that.

**Comment by daxpedda** [comment, marked as answer]
Unfortunately this is technically not possible. An `ArrayBuffer` represents external memory that can't be represented in pointers to the internal memory in Wasm (I don't believe the multi-memory proposal can fill that gap as well).

The solution is to accept a `Uint8Array` instead of an `&[u8]` to avoid having to copy everything into the Wasm memory, as long as that fits your requirements.

**Reply by DexterHill0** [reply]
That's unfortunate but makes sense. Not sure the solution is possible since the `image` library doesn't accept anything but a u8 slice. Am I correct in saying that if I wrapped the `UInt8Array` in an `ImageData`, passed that to the WASM, and then called `.data()` on the WASM side, that it would also have to allocated in WASM memory? I assume it does as `.data()` returns an owned vec but it's worth asking.

**Reply by daxpedda** [reply]
I assume the `UInt8Array` is pointing to external memory?

Currently [`ImageData::data()`](https://docs.rs/web-sys/0.3.67/web_sys/struct.ImageData.html#method.data) returns a `Vec<u8>`, so that definitely makes a copy. You would have to introduce custom bindings returning a [`Uint8ClampedArray`](https://docs.rs/js-sys/0.3.67/js_sys/struct.Uint8ClampedArray.html).

It is however possible to [construct `ImagaData` with internal memory through `Clamped`](https://docs.rs/web-sys/0.3.67/web_sys/struct.ImageData.html#method.new_with_u8_clamped_array). Same problem though: as `ImageData::data()` returns a `Vec`, allocation happens even though it's theoretically all pointing to Wasm memory. Not entirely sure if `wasm-bindgen` is able to generate bindings representing that though.

**Reply by DexterHill0** [reply]
> I assume the `UInt8Array` is pointing to external memory?
Yes, the array points to external memory. The array is constructed in JS and then passed to WASM.

Should have been clearer in my reply when I said "wrapped the `UInt8Array` in an `ImageData`". By this I mean wrap it in JS and pass the ImageData to the wasm. So create it `new ImageData(new UInt8ClampedArray(response.arraybuffer))` and then pass that value down to the WASM. But I think your point still applies that it would have to make a copy in the WASM.

**Reply by daxpedda** [reply]
Well you could still use a custom binding to `ImageData::data()` to returns a `Uint8ClampedArray` and do *something* with that without copying the data to Wasm memory. But if you want a `Vec` or `[u8]` you have to copy.

**Reply by DexterHill0** [reply]
Makes sense. Thanks for the help!
